core/templates/dev/head/pages/exploration_player/PlaythroughService.js

```diff
@@ -0,0 +1,275 @@
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value: multipleIncorrectStateName.state_name
+          },
+          num_times_answered_incorrectly: {
+            value: multipleIncorrectStateName.num_times_incorrect
+          }
+        };
+        return;
+      }
+
+      // Check for cyclic state transitions issue.
+      if (isCyclicStateTransitionsIssue()) {
+        playthrough.issueType = ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS;
+        playthrough.issueCustomizationArgs = {
+          state_names: {
+            value: cycleIdentifier.cycle.split(',')
+          }
+        };
+        return;
+      }
+
+      // Check for early quit issue.
+      if (isEarlyQuitIssue(timeSpentInExpInSecs)) {
+        playthrough.issueType = ISSUE_TYPE_EARLY_QUIT;
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value:
+              playthrough.actions[
+                playthrough.actions.length - 1].actionCustomizationArgs
+                .state_name.value
+          },
+          time_spent_in_exp_in_secs: {
+            value: expStopwatch.getTimeInSecs()
+          }
+        };
+        return;
+      }
+    };
+
+    var storePlaythrough = function(updatePlaythrough) {
+      if (updatePlaythrough) {
+        $http.post(getFullPlaythroughUrl(), {
+          playthrough_data: playthrough.convertToBackendDict(),
+          issue_schema_version: CURRENT_ISSUE_SCHEMA_VERSION
+        });
+      } else {
+        $http.post(getFullPlaythroughUrl(), {
+          playthrough_data: playthrough.convertToBackendDict(),
+          issue_schema_version: CURRENT_ISSUE_SCHEMA_VERSION
+        }).then(function(response) {
+          playthrough.playthroughId = response.playthroughId;
+        });
+      }
+    };
+
+    var getFullPlaythroughUrl = function() {
+      return UrlInterpolationService.interpolateUrl(
+        STORE_PLAYTHROUGH_URL, {
+          exploration_id: playthrough.expId
+        });
+    };
+
+    return {
+      initSession: function(newExplorationId, newExplorationVersion) {
+        playthrough = PlaythroughObjectFactory.createNew(
+          null, newExplorationId, newExplorationVersion, null, {}, []);
+        expStopwatch = StopwatchObjectFactory.create();
+      },
+      getPlaythrough: function() {
+        return playthrough;
+      },
+      recordExplorationStartAction: function(initStateName) {
+        playthrough.actions.push(LearnerActionObjectFactory.createNew(
+          ACTION_TYPE_EXPLORATION_START,
+          {
+            state_name: {
+              value: initStateName
+            }
+          },
+          CURRENT_ACTION_SCHEMA_VERSION
+        ));
+
+        multipleIncorrectStateName = {
+          state_name: initStateName,
+          num_times_incorrect: 0
+        };
+
+        cycleIdentifier = {
+          cycle: '',
+          num_cycles: 0
+        };
+        visitedStates.push(initStateName);
```

Comment:
I think these lines (199-208) would benefit from their own well-named function too (I had to double check to make sure these weren't a part of the learner object factory's backend dict, or some local variables).

Reply:
Done.